### PR TITLE
Feature/crud purchase

### PR DESCRIPTION
**Pull request by renatamarques97**

### CODE REVIEW

File: app/controllers/shopping_cart_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+      array << {
+        product: product, 
+        quantity: shopping_cart["quantity"]
+      }
+      array
+    end
+  end
+
+  def add_to_cart
```
Comment (renatamarques97) [inline app/controllers/shopping_cart_controller.rb:17]: too much logic in this controller @AleBL

File: app/models/product.rb [hunk 2]
```diff
@@ -3,4 +3,8 @@ class Product < ApplicationRecord
 
   validates :person_id, presence: true
   validates :person_type, presence: true
+
+  def dimensions
+    "#{height} cm x #{width} cm x #{length} cm (CxLxA)"
```
Comment (renatamarques97) [inline app/models/product.rb:8]: good catch!

File: config/routes.rb [hunk 3]
```diff
@@ -18,5 +16,13 @@
     password:      'adm/passwords'
   }
 
+  resources :products
+  resources :purchases
+
+  get  '/shopping_cart' => 'shopping_cart#index'
+  post '/add_to_cart/:product_id' => 'shopping_cart#add_to_cart', as: 'add_to_cart'
+  post '/remove_one_from_cart/:product_id' => 'shopping_cart#remove_one_from_cart', as: 'remove_one_from_cart'
```
Comment (renatamarques97) [inline config/routes.rb:24]: really good work here

File: spec/factories/products.rb [hunk 4]
```diff
@@ -1,11 +1,14 @@
...
-    weight      { FFaker::Random.rand(4.0) }
-    quantity    { FFaker::Random.rand(4) }
+    description { FFaker::Product.model }
+    height      { FFaker::Random.rand(10..80) }
+    width       { FFaker::Random.rand(10..80) }
+    length      { FFaker::Random.rand(10..80) }
+    weight      { FFaker::Random.rand(1.0..8.0).round(2) }
+    quantity    { FFaker::Random.rand(20..80) }
     person_type { "Person" }
+    person
```
Comment (renatamarques97) [inline spec/factories/products.rb:11]: you can remove the comments below

File: spec/models/purchase_spec.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Purchase, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
```
Comment (renatamarques97) [inline spec/models/purchase_spec.rb:4]: don't forget to add the specs

File: spec/views/products/edit.html.erb_spec.rb [hunk 6]
```diff
@@ -2,15 +2,7 @@
...
-    @product = assign(:product, Product.create!(
-      name: "MyString",
-      description: "MyText",
-      dimensions: "",
-      weight: 1.5,
-      quantity: 1,
-      person_type: "Person",
-      person_id: create(:person, :provider).id
-    ))
+    @product = assign(:product, create(:product))
```
Comment (renatamarques97) [inline spec/views/products/edit.html.erb_spec.rb:5]: so much better.

**Comment by renatamarques97** [comment]
please, see conflicts @AleBL

File: app/controllers/purchases_controller.rb [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    respond_to do |format|
+      format.html { redirect_to purchases_url, notice: 'Purchase was successfully destroyed.' }
+      format.json { head :no_content }
+    end
+  end
+  
+  private
+
+  def set_purchase
+    @purchase = Purchase.find(params[:id])
```
Comment (renatamarques97) [inline app/controllers/purchases_controller.rb:45]: please add "@_"

File: app/services/shopping_cart.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+      cart.reject { |item| item[:quantity].zero? }
+    end
+
+    def remove_all_from_cart(cart, product_id)
+      cart = convert_to_symbol(cart)
+      cart.reject { |item| item[:product_id] == product_id }
+    end
+
+    def cart_formatted(cart)
+      cart = convert_to_symbol(cart)
```
Comment (renatamarques97) [inline app/services/shopping_cart.rb:27]: you can extract this line, it's been using in four places.

File: app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -11,27 +11,47 @@
...
         <%= link_to "Sign up Client",   new_client_registration_path,   :class => 'navbar-link'  %> |
-        <%= link_to "Login Client",     new_client_session_path,        :class => 'navbar-link'  %>
+        <%= link_to "Login Client",     new_client_session_path,        :class => 'navbar-link'  %> |
         <%= link_to "Sign up Provider", new_provider_registration_path, :class => 'navbar-link'  %> |
         <%= link_to "Login Provider",   new_provider_session_path,      :class => 'navbar-link'  %>
       <% end %>
     </p>
-  </body>
+
+
```
Comment (renatamarques97) [inline app/views/layouts/application.html.erb:48]: remove ths extra line

File: spec/services/shopping_cart_spec.rb [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+          quantity: 3
+        },
+        {
+          product_id: products_id.third,
+          quantity: 6
+        }
+      ]
+    end
+
+    it "removed one from cart" do
```
Comment (renatamarques97) [inline spec/services/shopping_cart_spec.rb:79]: replace for "has been removed from cart"

File: spec/services/shopping_cart_spec.rb [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+      }]
+    end
+    let(:result_spec_already_added_to_cart) do
+      [{
+        product_id: product_id,
+        quantity: 2
+      }]
+    end
+
+    it "added to cart" do
```
Comment (renatamarques97) [inline spec/services/shopping_cart_spec.rb:31]: "add to cart"

File: spec/services/shopping_cart_spec.rb [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+          quantity: 2
+        },
+        {
+          product_id: products_id.second,
+          quantity: 4
+        }
+      ]
+    end
+
+    it "removed all from cart" do
```
Comment (renatamarques97) [inline spec/services/shopping_cart_spec.rb:119]: "remove ..."

**Review by Lucaas-Eduardo — APPROVED** [review]

**Review by AleBL — APPROVED** [review]

**Comment by AleBL** [comment]
[Criar opção de comprar (cliente)](https://trello.com/c/CYqzDonJ/24-criar-op%C3%A7%C3%A3o-de-comprar-cliente)

**Comment by AleBL** [comment]
[Criar carrinho](https://trello.com/c/FmIMMFID/12-criar-carrinho)

**Pull request merged**